### PR TITLE
[ENH] run GLM on surface data on BIDS datasets

### LINKED ISSUE #4567: Adapt first_level_from_bids to run on surface images directly

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Relates to https://github.com/nilearn/nilearn/issues/4027
- Closes #4567 

Changes proposed in this pull request:

- generate fake bids data with surface data
- adapt "first_level_from_bids" to look for surface data when space="fsaverage5" is requested.

### CODE REVIEW

**Comment by Remi-Gau** [comment]
Note that none of the bids datasets we use in the examples have surface data, so we can not demo this.

**Comment by Remi-Gau** [comment]
Failure in doc build indirectly reveals a bug: `first_level_from_bids` has no valid default when a space is not provided: will add a regression test and a fix.

```
     ../examples/08_experimental/plot_new_surface_bids_analysis_experimental.py failed leaving traceback:

    Traceback (most recent call last):
      File "/home/runner/work/nilearn/nilearn/examples/08_experimental/plot_new_surface_bids_analysis_experimental.py", line 72, in <module>
        models, run_imgs, events, confounds = first_level_from_bids(
                                              ^^^^^^^^^^^^^^^^^^^^^^
      File "/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/nilearn/glm/first_level/first_level.py", line 1608, in first_level_from_bids
        imgs, files_to_check = _get_processed_imgs(
                               ^^^^^^^^^^^^^^^^^^^^
      File "/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/nilearn/glm/first_level/first_level.py", line 1768, in _get_processed_imgs
        if space_label not in ("fsaverage5"):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    TypeError: 'in <string>' requires string as left operand, not NoneType
```

**Comment by Remi-Gau** [comment]
Trying on real data does not seem to work for now

```python
from pathlib import Path
from nilearn.glm.first_level import first_level_from_bids

data_dir = Path("/home/remi/datalad/datasets.datalad.org")
ds000001 = data_dir / "openneuro" / "ds000001"

derivatives_folder = str(Path("..") / ".." / "openneuro-derivatives" / "ds000001-fmriprep")
task_label = "balloonanalogrisktask"
space_label = "fsaverage5"

(
    models,
    models_run_imgs,
    models_events,
    models_confounds,
) = first_level_from_bids(
    dataset_path=ds000001,
    task_label=task_label,
    space_label=space_label,
    sub_labels=["01", "02"],
    t_r=None,
    slice_time_ref=None,
    mask_img=None,
    target_affine=None,
    target_shape=None,
    verbose=True,
    minimize_memory=True,
    derivatives_folder=derivatives_folder)
```


```
 /home/remi/miniconda3/bin/python /home/remi/github/nilearn/nilearn/tmp/load_from_fmriprep.py
/home/remi/github/nilearn/nilearn/tmp/load_from_fmriprep.py:16: UserWarning: 'StartTime' not found in file /home/remi/datalad/datasets.datalad.org/openneuro/ds000001/../../openneuro-derivatives/ds000001-fmriprep/sub-01/func/sub-01_task-balloonanalogrisktask_run-1_hemi-L_space-fsaverage5_bold.json.
  ) = first_level_from_bids(
/home/remi/github/nilearn/nilearn/nilearn/glm/first_level/first_level.py:1548: UserWarning: 'slice_time_ref' not provided and cannot be inferred from metadata.
It will be assumed that the slice timing reference is 0.0 percent of the repetition time.
If it is not the case it will need to be set manually in the generated list of models.
  warn(
Traceback (most recent call last):
  File "/home/remi/github/nilearn/nilearn/tmp/load_from_fmriprep.py", line 16, in <module>
    ) = first_level_from_bids(
        ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/remi/github/nilearn/nilearn/nilearn/glm/first_level/first_level.py", line 1610, in first_level_from_bids
    imgs, files_to_check = _get_processed_imgs(
                           ^^^^^^^^^^^^^^^^^^^^
  File "/home/remi/github/nilearn/nilearn/nilearn/glm/first_level/first_level.py", line 1819, in _get_processed_imgs
    SurfaceImage(
  File "/home/remi/github/nilearn/nilearn/nilearn/experimental/surface/_surface_image.py", line 221, in __init__
    _check_data_and_mesh_compat(self.mesh, self.data)
  File "/home/remi/github/nilearn/nilearn/nilearn/experimental/surface/_surface_image.py", line 181, in _check_data_and_mesh_compat
    raise ValueError(
ValueError: Data shape does not match number of vertices for 'right':
- data shape: (10242, 300)
- n vertices: 10242
```

File: nilearn/experimental/surface/_io.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 def read_array(array_file: Union[pathlib.Path, str]) -> np.ndarray:
     """Load surface data into a Numpy array."""
-    return old_surface.load_surf_data(array_file)
+    return old_surface.load_surf_data(array_file).T
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/experimental/surface/_io.py:14]: @jeromedockes 
was there any specific design choice for the dimensions of the data in the SurfaceImage object?

File: doc/changes/latest.rst [hunk 2]
```diff
@@ -19,6 +19,10 @@ Fixes
 
 - :bdg-dark:`Code` :func:`nilearn.image.binarize_img` explicitly cast images to ``int8`` to avoid warnings about ``int64`` when working with ``float64`` images (:gh:`4498` by `Patrick Sadil`_).
 
+- :bdg-dark:`Code` :func:`nilearn.glm.first_level.first_level_from_bids` willm for images in space ``- :bdg-dark:`Code` Improve :func:`nilearn.glm.first_level.first_level_from_bids` to look for and load ``fsaverage5`` data to easily run GLM on surface data (:gh:`4507` by `Rémi Gau`_).
```
Comment (Remi-Gau) [inline doc/changes/latest.rst:22]:
```suggestion
- :bdg-dark:`Code` :func:`nilearn.glm.first_level.first_level_from_bids` will for images in space 
```

File: doc/changes/latest.rst [hunk 3]
```diff
@@ -19,6 +19,10 @@ Fixes
 
 - :bdg-dark:`Code` :func:`nilearn.image.binarize_img` explicitly cast images to ``int8`` to avoid warnings about ``int64`` when working with ``float64`` images (:gh:`4498` by `Patrick Sadil`_).
 
+- :bdg-dark:`Code` :func:`nilearn.glm.first_level.first_level_from_bids` will for images in space 
+`` if not space is provided (:gh:`4507` by `Rémi Gau`_).
```
Comment (Remi-Gau) [inline doc/changes/latest.rst:23]:
```suggestion
if no space is provided (:gh:`4507` by `Rémi Gau`_).
```

File: doc/changes/latest.rst [hunk 4]
```diff
@@ -19,6 +19,10 @@ Fixes
 
 - :bdg-dark:`Code` :func:`nilearn.image.binarize_img` explicitly cast images to ``int8`` to avoid warnings about ``int64`` when working with ``float64`` images (:gh:`4498` by `Patrick Sadil`_).
 
+- :bdg-dark:`Code` :func:`nilearn.glm.first_level.first_level_from_bids` will for images in space
```
Comment (Remi-Gau) [inline doc/changes/latest.rst:22]:
```suggestion
- :bdg-dark:`Code` :func:`nilearn.glm.first_level.first_level_from_bids` will for images in MNI space 
```

File: nilearn/_utils/data_gen.py [hunk 5]
```diff
@@ -525,7 +525,7 @@ def write_fake_fmri_data_and_design(
     return mask_file, fmri_files, design_files
 
 
-def _write_fake_bold_gifti(file_path):
+def _write_fake_bold_gifti(file_path, shape, random_state=0):
     """Generate a gifti image and write it to disk.
 
     Note this only generates an empty file for now.
```
Comment (Remi-Gau) [inline nilearn/_utils/data_gen.py:531]:
```suggestion
    Note this only generates an empty file
    if the number of vertices demanded is 0.
```

File: nilearn/glm/tests/test_first_level.py [hunk 6]
```diff
@@ -1357,6 +1357,41 @@ def test_first_level_from_bids(
     assert len(imgs[0]) == n_imgs_expected
 
 
+@pytest.mark.parametrize("slice_time_ref", [None, 0.0, 0.5, 1.0])
+def test_first_level_from_bids_slice_time_ref(bids_dataset, slice_time_ref):
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:1361]: we had no test that covered those

File: nilearn/glm/tests/test_first_level.py [hunk 7]
```diff
@@ -1357,6 +1357,41 @@ def test_first_level_from_bids(
...
+        slice_time_ref=slice_time_ref,
+    )
+
+    _check_output_first_level_from_bids(n_sub, models, imgs, events, confounds)
+
+
+def test_first_level_from_bids_space_none(tmp_path):
+    """Test behavior when no specific space is required .
+
+    Function should look for images with NI152NLin2009cAsym.
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:1378]:
```suggestion
    Function should look for images with MNI152NLin2009cAsym.
```

File: nilearn/experimental/surface/_io.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/experimental/surface/_io.py:14]: skikit_learn has n_samples * n_features

what would nibabel do ?

[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/experimental/surface/_io.py:14]: regarding nibabel it stores gifti data in a list of n GiftiDataArray (n == nb timepoints), where each GiftiDataArray has m vertices.

so in a way the first dimension is the number of timepoints.

```python
from pathlib import Path
import nibabel as nib
from rich import print

dir = Path("/home/remi/datalad/datasets.datalad.org/openneuro-derivatives/ds003500-fmriprep/sub-221/func")

file = dir / "sub-221_task-Conj19Sel_acq-b_hemi-L_space-fsaverage5_bold.func.gii"

gii = nib.load(file)
print(gii.__dict__)
print(type(gii.darrays))
print(len(gii.darrays))
```

```
{
    '_header': <nibabel.filebasedimages.FileBasedHeader object at 0x7ad960d46990>,
    'extra': {},
    'file_map': {'image': <nibabel.fileholders.FileHolder object at 0x7ad960d469c0>},
    '_labeltable': <GiftiLabelTable []>,
    '_meta': <GiftiMetaData {'UserName': 'jbwexler', 'Date': 'Thu Oct 13 22:21:12 2022', 'gifticlib-version': 'gifti library version 
1.09, 28 June, 2010', 'AnatomicalStructurePrimary': 'CortexLeft'}>,
    'darrays': [
        <GiftiDataArray time series[10242]>,
        <GiftiDataArray time series[10242]>,
       ....
        <GiftiDataArray time series[10242]>,
        <GiftiDataArray time series[10242]>
    ],
    'version': '1.0'
}
<class 'list'>
156
```

[the remark this answers is not in the document]
Comment (bthirion) [reply nilearn/experimental/surface/_io.py:14]: So n_time points * n_vertices definitely makes sense

[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/experimental/surface/_io.py:14]: yup.

will open a small PR to update this when I work on adding a feature to save the data of a SurfaceImage to disk (right now we can only save meshes)

**Comment by Remi-Gau** [comment]
Now we can easily load data from fmriprep outpout

```python
from pathlib import Path

from nilearn.surface import SurfaceImage
from nilearn.datasets import load_fsaverage

dir = Path("/home/remi/datalad/datasets.datalad.org/openneuro-derivatives/ds003500-fmriprep/sub-221/func")

file_L = dir / "sub-221_task-Conj19Sel_acq-b_hemi-L_space-fsaverage5_bold.func.gii"
file_R = dir / "sub-221_task-Conj19Sel_acq-b_hemi-R_space-fsaverage5_bold.func.gii"

img = SurfaceImage(mesh=load_fsaverage()["pial"], data={"left":file_L, "right":file_R})
print(img)
```

File: nilearn/glm/first_level/first_level.py [hunk 8]
```diff
@@ -1764,9 +1771,15 @@ def _get_processed_imgs(
 
     Returns
     -------
-    imgs : :obj:`list` of :obj:`str`
+    imgs : :obj:`list` of :obj:`str`, or :obj:`list` of :obj:`SurfaceImage`
```
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:1774]:
```suggestion
    imgs : :obj:`list` of :obj:`str`, or :obj:`list` of :obj:`~nilearn.surface.SurfaceImage`
```

**Comment by Remi-Gau** [comment]
could easily run a GLM on a random openneuro dataset

```python
from pathlib import Path

import numpy as np
from nilearn.datasets import load_fsaverage, load_fsaverage_data
from nilearn.glm.first_level import first_level_from_bids
from nilearn.plotting import plot_surf_stat_map, show

dataset_path = Path("/home/remi/datalad/datasets.datalad.org/openneuro/ds003500")


models, run_imgs, events, confounds = first_level_from_bids(
    dataset_path=dataset_path,
    task_label="Conj19Sel",
    space_label="fsaverage5",
    sub_labels=["221"],
    img_filters=None,
    derivatives_folder=str(
        dataset_path / ".." / ".." / "openneuro-derivatives" / "ds003500-fmriprep"
    ),
)

for first_level_glm, image, confound, event in zip(models, run_imgs, confounds, events):
    first_level_glm.fit(run_imgs=image, events=event)
    result = first_level_glm.compute_contrast(contrast_def=np.ones(1))
    plot_surf_stat_map(
        surf_mesh=load_fsaverage().inflated,
        stat_map=result,
        threshold=1.96,
        bg_map=load_fsaverage_data(mesh_type="inflated"),
        cmap="bwr",
    )
    show()
```

Figure_1

**Comment by Remi-Gau** [comment]
@bthirion this one is out of the scope for the 0.11 release but I think it is ready to go anyway

**Review by man-shu — APPROVED** [review]
LGTM otherwise!

File: nilearn/_utils/data_gen.py [hunk 9]
```diff
@@ -537,9 +538,25 @@ def _write_fake_bold_gifti(file_path):
...
+    random_state : :obj:`int` or :obj:`numpy.random.RandomState` instance, \
+                   default=0
+        Random number generator, or seed.
     """
-    Path(file_path).parent.mkdir(parents=True, exist_ok=True)
-    Path(file_path).touch()
+    if shape[0] == 0:
+        Path(file_path).parent.mkdir(parents=True, exist_ok=True)
+        Path(file_path).touch()
+        return file_path
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/_utils/data_gen.py:552]: ok checked locally and those lines are not covered indeed

**Pull request merged**